### DISCUSSION TITLE
Artifactory rubygems bug

### DISCUSSION THREAD

**Discussion by 15joeybloom** [renovatebot/renovate#12497] [Request Help]
I have found a bug that causes renovate to miss some dependency updates.

I've configured renovate (cloud hosted, version 28.21.0) on a ruby project with our artifactory rubygems repository as a datasource. I have the `license_finder` gem in our `Gemfile` with the constraint `~> 6.0.0` and locked in our `Gemfile.lock` at `6.0.0`. Artifactory reports the latest version as `6.14.2`.

I'd expect renovate to open a PR updating the `Gemfile` to `~> 6.14.0` and `Gemfile.lock` to `6.14.2`. Instead, renovate totally misses the update.

I've narrowed down the problem to the `getCurrentVersion` function: https://github.com/renovatebot/renovate/blob/4e24979f9fc06cd2a9bd06480c7852d0659e6f33/lib/workers/repository/process/lookup/current.ts#L7-L14 Our `currentValue` is `~> 6.0.0`, `lockedVersion` is `6.0.0`, and `allVersions` is `["6.14.2"]`. The function returns `"6.14.2"` as the "current version", which causes renovate to skip the update because it thinks that we're on the latest version already. But we'd expect `getCurrentVersion` to instead return `"6.0.0"`; that is the version our project is currently using because that's what's in the lockfile.

There's a [known issue](https://github.com/renovatebot/renovate/issues/7320) with artifactory rubygems repositories: Artifactory will only tell renovate about the latest version of a gem; there's no way to find out about older versions. That's why `allVersions` is `["6.14.2"]` in this case and so it's partially causing this issue. If `allVersions` also included `"6.0.0"` then `getCurrentVersion` would return `"6.0.0"` as expected.

A potential fix for this issue is to treat `lockedVersion` as the current version if it's provided. It seems like a safe assumption to me. To verify this, I ran renovate from my laptop with the following added to the beginning of `getCurrentVersion`:
```ts
  if (versioning.isVersion(lockedVersion)) {
    return lockedVersion
  }
```
As expected, renovate created a PR to update `license_finder` to `6.14.2`. I also ran `yarn test` and it passed all unit tests.

I'd like to discuss integrating this fix into renovate. How should I proceed? I can't really provide an example github repo to reproduce the issue since it only happens with an artifactory rubygems repository that I can't share.

**Comment by rarkins** [comment]
Please see if an alternative rangeStrategy helps

**Reply by 15joeybloom** [reply]
Tried all the rangeStrategies (auto,pin,bump,replace,widen,update-lockfile) and the only one that resulted in a license_finder update was `"pin"`. This makes sense as this is the only case in which `getCurrentVersion` even considers the `lockedVersion`. 

However I don't want to change range strategies. I've been happy with the default "replace" strategy so far.

The fix I'm proposing would work for all range strategies.

**Reply by rarkins** [reply]
What you proposed is not correct behaviour for "replace"

**Reply by 15joeybloom** [reply]
To make sure we're talking about the same thing, the "replace" rangeStrategy is "Replace the range with a newer one if the new version falls outside it, and update nothing otherwise"

In my case the range is `~> 6.0.0`, and the new version is `6.14.2`. The new version falls outside the range, so I would expect the range to be replaced with a newer one, probably `~> 6.14.0`. 

In the original post I was actually using the "replace" rangeStrategy since it's the default, so I don't see how my solution breaks it. Can you explain why?

**Reply by rarkins** [reply]
Sorry, you're right. If current value is `~> 6.0.0` and the version is `6.14.2` then you'd expect `~> 6.14.0`.

But the root cause here is that Artifactory has an incomplete Rubygems implementation, so what's needed is to work around the Artifactory problem. I don't agree with the idea that Renovate has a "bug" because it can't currently work around a broken Artifactory result, unless it happens to be coincidental.

You mention that `getCurrentVersion()` returns `6.14.2` which I find surprising, because `6.14.2` should not satisfy `~> 6.0.0`. Which line of that function is returning the value?

**Reply by 15joeybloom** [reply]
Sorry, it's my turn to make a mistake. I saw `getCurrentVersion` returning `6.14.2` only when currentValue was `~> 6.0`. When currentValue is instead `~> 6.0.0`, `getCurrentVersion` returns `null` because `6.14.2` doesn't satisfy `~> 6.0.0`.

In both cases, the locked version `6.0.0` is ignored. I think it would be better for `getCurrentVersion` to use the locked version and return `6.0.0` than to return `null` or `6.14.2`. Do you agree?

I agree that there's only so much renovate can do if artifactory won't support all the necessary APIs. I think this is just a coincidental issue.

**Reply by rarkins** [reply]
Keep in mind that Renovate's default assumption is that if a registry says a version doesn't exist, then we assume it doesn't exist, including that it's been yanked due to being malicious.

I don't think there's a need to get into `getCurrentVersion` specifics here, because it's an internal temporary variable, maybe even poorly named. Ultimately you just want to see a PR with `~> 6.14.0`, right? That can possibly be done regardless of what getCurrentVersion returns.

**Reply by 15joeybloom** [reply]
> Ultimately you just want to see a PR with ~> 6.14.0, right? That can possibly be done regardless of what getCurrentVersion returns.

Yes, that is what we'd like to happen. To reiterate,

1. We specify `~> 6.0.0` in our Gemfile and pin `6.0.0` in Gemfile.lock
2. Our registry reports `6.14.2` as the latest version
3. Renovate should create a PR to upgrade to the newer version. `~> 6.14.0` in the Gemfile would make sense.

Does that make sense to you? If so, we would love to contribute a PR if you can recommend where to make changes.

**Reply by rarkins** [reply]
A PR would be welcome, but I have to caution you that editing the lookup code comes with a [Here be dragons](https://en.wikipedia.org/wiki/Here_be_dragons) warning. As in, we need to be really careful not to make unintended changes to other ones (especially because this is really an edge case workaround for a poor registry). That said, it would be best if Renovate proposes the update you're expecting with rangeStrategy=replace

**Reply by eigenhombre** [reply]
@rarkins We are working on a PR.  Can/should this be converted to an Issue that we reference in our commits?  Thanks in advance.

**Reply by rarkins** [reply]
See #12618
